py3status/modules/playerctl.py

```diff
@@ -0,0 +1,268 @@
+    color_paused: Paused, defaults to color_degraded
+    color_playing: Playing, defaults to color_good
+    color_stopped: Stopped, defaults to color_bad
+
+Requires:
+    playerctl: mpris media player controller and lib for spotify, vlc, audacious,
+        bmp, xmms2, and others.
+
+@author jdholtz
+
+SAMPLE OUTPUT
+[
+    {'color': '#00FF00', 'full_text': '> My Lucky Pants Failed Me Again - Tom Rosenthal'}
+    {'color': '#FFFF00', 'full_text': '|| Too Much Skunk Tonight - Birdy Nam Nam'}
+]
+"""
+from threading import Thread
+
+import gi
+
+gi.require_version("Playerctl", "2.0")
+from gi.repository import GLib, Playerctl
+
+
+class Py3status:
+    """ """
+
+    # available configuration parameters
+    button_next = None
+    button_pause = None
+    button_play = None
+    button_play_pause = 1
+    button_previous = None
+    button_stop = 3
+    cache_timeout = 5
+    format = "{format_player}"
+    format_player = (
+        r"[\?if=is_playing > ][\?if=is_paused \|\| ]"
+        r"[\?if=is_stopped .. ][[{artist}][\?soft  - ][{title}]]"
+    )
+    format_separator = " "
+    players = []
+    sleep_timeout = 20
+
+    def post_config_hook(self):
+        if not self.py3.check_commands("playerctl"):
+            raise Exception("command 'playerctl' not installed")
+
+        self.color_paused = self.py3.COLOR_PAUSED or self.py3.COLOR_DEGRADED
+        self.color_playing = self.py3.COLOR_PLAYING or self.py3.COLOR_GOOD
+        self.color_stopped = self.py3.COLOR_STOPPED or self.py3.COLOR_BAD
+
+        self._init_manager()
+
+    def _init_manager(self):
+        self.manager = Playerctl.PlayerManager()
+        self.manager.connect("name-appeared", self._on_name_appeared)
+        self.manager.connect("name-vanished", self._on_name_vanished)
+
+        # Initialize all currently active players
+        for player_name in self.manager.props.player_names:
+            self._init_player(player_name)
+
+        self._start_loop()
+
+    def _init_player(self, player_name):
+        if len(self.players) > 0 and player_name.name not in self.players:
+            return
+
+        player = Playerctl.Player.new_from_name(player_name)
+
+        # Set up all event listeners
+        player.connect("playback-status", self._status_changed, self.manager)
+        player.connect("metadata", self._on_metadata, self.manager)
+
+        self.manager.manage_player(player)
+
+    def _start_loop(self):
+        loop = GLib.MainLoop()
+
+        # The loop is blocking so it needs to be run a separate thread
+        thread = Thread(target=loop.run)
+        thread.daemon = True
+        thread.start()
+
+    def _on_name_appeared(self, manager, player_name):
+        self._init_player(player_name)
+
+    def _on_name_vanished(self, manager, name):
+        self.py3.update()
+
+    def _on_metadata(self, player, metadata, manager):
+        self.py3.update()
+
+    def _status_changed(self, player, status, manager):
+        self.py3.update()
```

Comment:
About timing... Using playerctl command, it would make sense... With python bindings,  we shouldn't need `cache_timeout` and `sleep_timeout` config anymore. 

Fix to return  `{"cached_until": self.py3.CACHE_FOREVER, ...` instead.

Reply:
The refresh is there to update the position in the song/video (which is a valid attribute). Otherwise, it wouldn't update at all. What I can do is add a timeout only if the position is a placeholder and then set a timeout based on that (could be done in the `post_config_hook`).

Reply:
Is this confirmed? So far, it looks like this module is being updated consistently even with `self.py3.CACHE_FOREVER` tossed in there due to metadata/playback-status (possibly on 3 players).

Reply:
It updates whenever the metadata or playback status changes, but the `position` placeholder is also offered in this module. When I add that to the `format_player` and use `"cached_until": self.py3.CACHE_FOREVER`, the position does not update unless the metadata/status changes.

Reply:
That sounds so gooood. We might not need cache_timeout after all too. If position is included, just hardcode it to a second, then when players aren't tracked, cache_forever again.